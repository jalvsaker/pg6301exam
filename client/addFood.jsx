import { useState } from "react";

export function AddFood({ reload }) {
  const [name, setName] = useState("");
  const [price, setPrice] = useState("");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price };

    await fetch("/api/foods", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    reload();
  }

  return (
    <form onSubmit={handleSubmit}>
      <label>
        Name:
        <input
          type={"text"}
          value={name}
          onChange={(e) => setName(e.target.value)}
        />
      </label>
      <label>
        Price:
        <input
          type={"number"}
          value={price}
          onChange={(e) => setPrice(e.target.value)}
        />
      </label>
      <button>Submit</button>
    </form>
  );
}
