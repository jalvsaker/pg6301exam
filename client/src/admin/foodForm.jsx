import React from "react";
import { useState } from "react";

export function FoodForm({ onSubmit, food, buttonText, resetAfter }) {
  const [name, setName] = useState(food?.name || "");
  const [price, setPrice] = useState(food?.price || "");
  const [description, setDescription] = useState(food?.description || "");

  async function handleSubmit(e) {
    e.preventDefault();

    const body = { name, price, description };

    await onSubmit(body);

    if (resetAfter) {
      setName("");
      setPrice("");
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      <div>
        <label>
          Name:
          <input
            type={"text"}
            value={name}
            onChange={(e) => setName(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Price:
          <input
            type={"number"}
            value={price}
            onChange={(e) => setPrice(e.target.value)}
          />
        </label>
      </div>
      <div>
        <label>
          Description:
          <textarea
            cols="30"
            rows="4"
            value={description}
            onChange={(e) => setDescription(e.target.value)}
          ></textarea>
        </label>
      </div>
      <button>{buttonText}</button>
    </form>
  );
}
