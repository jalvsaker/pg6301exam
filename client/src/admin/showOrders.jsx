import React, { useEffect, useState } from "react";
import { Link, useNavigate } from "react-router-dom";

export function ShowOrders({user}) {
  const [orders, setOrders] = useState([]);
  const [loading, setLoading] = useState(true);

  const navigate = useNavigate();

  useEffect(() => {
    if (!user.isAdmin) {
      navigate("/");
    }
  }, []);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/orders");

      setOrders(await res.json());
      setLoading(false);
    })();
  }, []);

  if (loading) {
    return "Loading...";
  }

  return <>
    <h2>Orders:</h2>
    <Link to={"/admin"}>&lt;-Back</Link>
    {orders.map((order) => <div key={order._id}>
      <h4>Order by user {order.username} to be delivered at {order.time}</h4>
      <h5>Total price: {order.totalPrice.toFixed(2)}</h5>
      <ul>
        {order.cart.map((item) => <li key={item.name} >{item.amount} {item.name} {item.price}</li>)}
      </ul>
      <hr/>
    </div>)}
  </>;
}
