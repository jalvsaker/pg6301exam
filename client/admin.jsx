import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <AddFood reload={reload} />

      {foods.map((food) => (
        <div key={food._id}>{food.name}</div>
      ))}
    </>
  );
}
