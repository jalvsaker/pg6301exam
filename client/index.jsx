import React from "react";
import { createRoot } from "react-dom/client";
import {
  BrowserRouter,
  Navigate,
  NavLink,
  Route,
  Routes,
} from "react-router-dom";
import { Homepage } from "./homepage";

function App() {
  return (
    <BrowserRouter>
      <nav>
        <NavLink to={"/"}>Home</NavLink>
        <NavLink to={"/test"}>Test</NavLink>
      </nav>
      <Routes>
        <Route path={"/"} element={<Homepage />} />
        <Route path={"/test"} element={<h2>test</h2>} />
        <Route path={"/*"} element={<Navigate to={"/"} />} />
      </Routes>
    </BrowserRouter>
  );
}

const root = createRoot(document.getElementById("root"));
root.render(
  <React.StrictMode>
    <App />
  </React.StrictMode>
);
