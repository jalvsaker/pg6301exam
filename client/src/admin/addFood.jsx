import { FoodForm } from "./foodForm";

export function AddFood({reload}){

  async function submit(body) {
    await fetch("/api/foods", {
      method: "POST",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body)
    });

    reload();
  }


  return <FoodForm onSubmit={submit} buttonText={"Add"} resetAfter={true}/>
}

