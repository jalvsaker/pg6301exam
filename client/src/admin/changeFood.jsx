import { FoodForm } from "./foodForm";
import { useState } from "react";

export function ChangeFood({ food, reload }) {
  const [success, setSuccess] = useState(false);

  async function submit(body) {
    const res = await fetch(`/api/foods/${food._id}`, {
      method: "PUT",
      headers: { "content-type": "application/json" },
      body: JSON.stringify(body),
    });

    if (res.ok) {
      setSuccess(true);
    }

    reload();
  }
  async function deleteFood(e) {
    e.preventDefault();

    await fetch(`/api/foods/${food._id}`, {
      method: "DELETE",
    });

    reload();
  }

  return (
    <>
      {success && <h4>Updated</h4>}
      <FoodForm
        food={food}
        onSubmit={submit}
        buttonText={"Update"}
        resetAfter={false}
      />
      <button onClick={deleteFood}>Delete</button>
      <hr />
    </>
  );
}
