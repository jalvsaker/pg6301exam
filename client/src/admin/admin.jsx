import { useNavigate } from "react-router-dom";
import { useEffect, useState } from "react";
import { AddFood } from "./addFood";
import { ChangeFood } from "./changeFood";

export function Admin({ user }) {
  const navigate = useNavigate();

  if (!user.isAdmin) {
    navigate("/");
  }

  const [foods, setFoods] = useState([]);
  const [loading, setLoading] = useState(true);
  const [reloader, setReloader] = useState(false);

  useEffect(() => {
    (async () => {
      const res = await fetch("/api/foods");
      if (res.ok) {
        setFoods(await res.json());
        setLoading(false);
      }
    })();
  }, [reloader]);

  function reload() {
    setReloader((r) => !r);
  }

  if (loading) {
    return <div>Loading...</div>;
  }

  return (
    <>
      <h1>Admin panel</h1>

      <h3>Add new foods:</h3>
      <AddFood reload={reload} />

      <h3>Change previous foods:</h3>

      {foods.map((food) => (
        <ChangeFood key={food._id} food={food} reload={reload}/>
      ))}
    </>
  );
}
