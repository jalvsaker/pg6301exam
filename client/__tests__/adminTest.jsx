import React from "react";
import { createRoot } from "react-dom/client";
import { act, Simulate } from "react-dom/test-utils";
import { AddFood } from "../src/admin/addFood";
import { ChangeFood } from "../src/admin/changeFood";
import { MemoryRouter } from "react-router-dom";
import { Admin } from "../src/admin/admin";
import { FoodForm } from "../src/admin/foodForm";
import { ShowOrders } from "../src/admin/showOrders";

describe("admin tests", function () {
  it("should show addFood", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: false,
        json: () => {
          return [];
        },
      };
    });

    const mock = jest.fn()

    await act(() => {
      root.render(<AddFood reload={mock} />);
    });

    await act(()=> {
      Simulate.submit(element.querySelector("form"))
    })

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toHaveBeenCalled();
  });

  it("should show changeFood", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: false,
        json: () => {
          return [];
        },
      };
    });

    const mock = jest.fn()

    await act(() => {
      root.render(<ChangeFood food={{_id: 1}} reload={mock}/>);
    });

    await act(()=> {
      Simulate.submit(element.querySelector("form"))
    })

    await act(() => {
      Simulate.click(element.querySelectorAll("button")[1])
    })

    expect(element.innerHTML).toMatchSnapshot();
    expect(global.fetch).toHaveBeenCalled();
  });

  it("should show Admin panel", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [];
        },
      };
    });

    await act(() => {
      root.render(
        <MemoryRouter>
          <Admin user={{ isAdmin: true }} />
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should show not show Admin panel because user is not admin", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [];
        },
      };
    });

    await act(() => {
      root.render(
        <MemoryRouter>
          <Admin user={{ isAdmin: false }} />
        </MemoryRouter>
      );
    });

    expect(element.innerHTML).toMatchSnapshot();
  });

  it("should submit food form", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    const mock = jest.fn();

    await act(() => {
      root.render(<FoodForm onSubmit={mock} />);
    });

    await act(() => {
      Simulate.change(element.getElementsByTagName("input")[0], {
        target: { value: "foodname" },
      });
      Simulate.change(element.getElementsByTagName("input")[1], {
        target: { value: "12" },
      });
    });

    await act(() => {
      Simulate.submit(element.querySelector("form"));
    });

    expect(element.innerHTML).toMatchSnapshot();
    expect(mock).toBeCalledWith({ name: "foodname", price: "12" });
  });

  it("should show orders", async () => {
    const element = document.createElement("div");
    const root = createRoot(element);

    global.fetch = jest.fn(() => {
      return {
        ok: true,
        json: () => {
          return [
            {
              "_id": "1",
              "username": "hei",
              "totalPrice": 41,
              "time": "12:32",
              "cart": [
                {
                  "name": "Burgers",
                  "price": "8.2",
                  "amount": "5"
                }
              ]
            },
            {
              "_id": "2",
              "username": "hallo",
              "totalPrice": 42.6,
              "time": "12:59",
              "cart": [
                {
                  "name": "Burgers",
                  "price": "8.2",
                  "amount": 3
                },
                {
                  "name": "Pizza",
                  "price": "6",
                  "amount": 3
                }
              ]
            }
          ];
        },
      };
    });

    await act(() => {
      root.render(<MemoryRouter><ShowOrders user={{ isAdmin: true }} /></MemoryRouter>);
    });

    expect(element.innerHTML).toMatchSnapshot();
  });
});
