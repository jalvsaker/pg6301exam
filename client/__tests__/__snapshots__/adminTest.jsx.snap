// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`admin tests should show Admin panel 1`] = `"<h1>Admin panel</h1><a href="/admin/orders">Show orders</a><hr><h3>Add new foods:</h3><form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><div><label>Description:<textarea cols="30" rows="4"></textarea></label></div><button>Add</button></form><h3>Change previous foods:</h3>"`;

exports[`admin tests should show addFood 1`] = `"<form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><div><label>Description:<textarea cols="30" rows="4"></textarea></label></div><button>Add</button></form>"`;

exports[`admin tests should show changeFood 1`] = `"<form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><div><label>Description:<textarea cols="30" rows="4"></textarea></label></div><button>Update</button></form><button>Delete</button><hr>"`;

exports[`admin tests should show not show Admin panel because user is not admin 1`] = `"<h1>Admin panel</h1><a href="/admin/orders">Show orders</a><hr><h3>Add new foods:</h3><form><div><label>Name:<input type="text" value=""></label></div><div><label>Price:<input type="number" value=""></label></div><div><label>Description:<textarea cols="30" rows="4"></textarea></label></div><button>Add</button></form><h3>Change previous foods:</h3>"`;

exports[`admin tests should show orders 1`] = `"<h2>Orders:</h2><a href="/admin">&lt;-Back</a><div><h4>Order by user hei to be delivered at 12:32</h4><h5>Total price: $41.00</h5><ul><li>5 Burgers $8.2</li></ul><hr></div><div><h4>Order by user hallo to be delivered at 12:59</h4><h5>Total price: $42.60</h5><ul><li>3 Burgers $8.2</li><li>3 Pizza $6</li></ul><hr></div>"`;

exports[`admin tests should submit food form 1`] = `"<form><div><label>Name:<input type="text" value="foodname"></label></div><div><label>Price:<input type="number" value="12"></label></div><div><label>Description:<textarea cols="30" rows="4"></textarea></label></div><button></button></form>"`;
